### PR TITLE
[Feature] Diagnostic infrastructure

### PR DESCRIPTION
**Pull request by lquerel**
# Phase 1
This PR is the first in a series of two, aiming to separate errors from the concept of diagnostic messages (see #124). An error causes the program to stop, whereas diagnostic messages are collected and displayed progressively to allow users to make the necessary corrections. This approach is similar to the typical operation of a compiler. The ultimate goal is to improve the user experience by providing useful, nice, and fast diagnostic messages.

**Why use an MPSC channel in the implementation?**
Weaver is parallelized at multiple levels (e.g., semconv parsing, template rendering, policy enforcement). Each thread can emit diagnostic messages, and the current implementation requires an Arc<Mutex<Logger>> plus clones for the logger. Furthermore, the `paris` logging library we currently use does not allow calling stdout/err().lock() for exclusive access to the console for fast outputs. The combination of these two factors makes the rendering of error messages relatively slow. One way to solve these problems is to use an MPSC channel to allow a dedicated thread to handle all console outputs in an optimized manner.

With this channel concept, there is no longer a need for a `CompoundError` because diagnostics are reported progressively through the channel (see the example and unit tests provided for more details).

**Example of output**
Screenshot demonstrating the possibilities of this infrastructure. The crate `miette` and its derive macro is used to specify rich diagnostic messages (message, help, note, url, code span, ...).

image

# Phase 2
The integration of this infrastructure with the rest of the Weaver project will done in a second PR.

### CODE REVIEW

File: crates/weaver_common/src/diag/consumer/console.rs [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+use std::sync::mpsc;
+
+/// A consumer that writes diagnostic messages to the console as plain text.
+pub struct ConsoleDiagMessageConsumer {
+    stdout_lock: bool,
+}
+
+impl ConsoleDiagMessageConsumer {
+    /// Creates a new console consumer.
+    pub fn new(stdout_lock: bool) -> Self {
```
[the remark this answers is not in the document]
Comment (lquerel) [reply crates/weaver_common/src/diag/consumer/console.rs:16]: Fixed

File: crates/weaver_common/src/diag/consumer/console.rs [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+            None
+        };
+
+        for message in receiver {
+            match message {
+                SystemMessage::Diagnostic(report) => {
+                    if let Some(severity) = report.severity() {
+                        match severity {
+                            miette::Severity::Advice => {
+                                println!("Advice: {:?}", report);
```
[the remark this answers is not in the document]
Comment (lquerel) [reply crates/weaver_common/src/diag/consumer/console.rs:40]: Fixed

File: crates/weaver_common/src/diag/consumer/console.rs [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+                            }
+                            miette::Severity::Warning => {
+                                eprintln!("Warning: {:?}", report);
+                            }
+                            miette::Severity::Error => {
+                                eprintln!("Error: {:?}", report);
```
[the remark this answers is not in the document]
Comment (lquerel) [reply crates/weaver_common/src/diag/consumer/console.rs:46]: Fixed

File: crates/weaver_common/src/diag/consumer/console.rs [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+                            }
+                        }
+                    } else {
+                        println!("{:?}", report);
```
[the remark this answers is not in the document]
Comment (lquerel) [reply crates/weaver_common/src/diag/consumer/console.rs:50]: Fixed

**Review by jsuereth — APPROVED** [review]
As discussed offline -

- I like miette a lot, we should absolutely use this going forward.
- I'm skeptical on Diagnostic Channel from past experience.  I think we may need more state / bidirectional communication of warnings + errors.


Approving this one two conditions:

1. We agree not to use `DiagChannel` in weaver until further prototyping occurs.
2. The two of us spend some time prototyping alternative mechanisms of propagating diagnostic message *and* errors through weaver to get a feel for how this plays out.

I suspect this is one of the most important decisions in weaver architecture so I don't want to rush into it without some exploration into what's possible today that wasn't in reference architectures.

**Comment by lquerel** [comment]
@jsuereth I agree, it is important to get this infrastructure right. I will create a new PR to explore the DiagChannel approach, so we can review it in parallel with your PR.

**Pull request merged**